tests/frame_manager_test.go

```diff
@@ -33,37 +35,79 @@ func TestWaitForFrameNavigationWithinDocument(t *testing.T) {
 	for _, tc := range testCases {
 		tc := tc
 		t.Run(tc.name, func(t *testing.T) {
-			tb := newTestBrowser(t, withFileServer())
-			p := tb.NewPage(nil)
+			t.Parallel()
 
-			resp := p.Goto(tb.staticURL("/nav_in_doc.html"), nil)
-			require.NotNil(t, resp)
+			errc := make(chan error, 1)
+			go func() {
+				tb := newTestBrowser(t, withFileServer())
+				p := tb.NewPage(nil)
 
-			// A click right away could possibly trigger navigation before we
-			// had a chance to call WaitForNavigation below, so give it some
-			// time to simulate the JS overhead, waiting for XHR response, etc.
-			time.AfterFunc(timeout*time.Millisecond, func() { //nolint:durationcheck
-				_ = tb.await(func() error {
-					_ = p.Click(tc.selector, nil)
-					return nil
-				})
-			})
+				resp := p.Goto(tb.staticURL("/nav_in_doc.html"), nil)
+				require.NotNil(t, resp)
 
-			done := make(chan struct{}, 1)
-			go func() {
-				require.NotPanics(t, func() {
-					p.WaitForNavigation(tb.toGojaValue(&common.FrameWaitForNavigationOptions{
-						Timeout: timeout * 3, // interpreted as ms
+				// A click right away could possibly trigger navigation before we
+				// had a chance to call WaitForNavigation below, so give it some
+				// time to simulate the JS overhead, waiting for XHR
+				// response, etc.
+				<-time.After(timeout * time.Millisecond) //nolint:durationcheck
+
+				// if one of the promises panics, err will contain the error
+				errc <- tb.await(func() error {
+					_ = p.Click(tc.selector, nil)
+					_ = p.WaitForNavigation(tb.toGojaValue(&common.FrameWaitForNavigationOptions{
+						Timeout: 3 * timeout, // interpreted as ms
 					}))
+					return nil
 				})
-				done <- struct{}{}
 			}()
-
 			select {
-			case <-done:
-			case <-time.After(timeout * 5 * time.Millisecond): //nolint:durationcheck
+			case err := <-errc:
+				require.NoError(t, err)
+			case <-time.After(5 * timeout * time.Millisecond): //nolint:durationcheck
 				t.Fatal("Test timed out")
 			}
 		})
 	}
 }
+
+func TestWaitForFrameNavigation(t *testing.T) {
+	tb := newTestBrowser(t, withHTTPServer())
+	p := tb.NewPage(nil)
+
+	tb.withHandler("/first", func(w http.ResponseWriter, _ *http.Request) {
+		fmt.Fprintf(w, `
+		<html>
+			<head>
+				<title>First page</title>
+			</head>
+			<body>
+				<a href="/second">click me</a>
+			</body>
+		</html>
+		`)
+	})
+	tb.withHandler("/second", func(w http.ResponseWriter, _ *http.Request) {
+		fmt.Fprintf(w, `
+		<html>
+			<head>
+				<title>Second page</title>
+			</head>
+			<body>
+				<a href="/first">click me</a>
+			</body>
+		</html>
+		`)
+	})
+
+	require.NotNil(t, p.Goto(tb.URL("/first"), tb.toGojaValue(&common.FrameGotoOptions{
+		WaitUntil: common.LifecycleEventNetworkIdle,
+		Timeout:   common.DefaultTimeout,
+	})))
+	err := tb.await(func() error {
+		_ = p.Click(`a`, nil)
+		p.WaitForNavigation(nil)
+		return nil
+	})
```

Comment:
Same here, I don't see how these Promises are resolved or synchronized, or that `tb.await()` (`vu.RunLoop()`) actually waits for them to run. In fact, this test passes if you remove the `p.WaitForNavigation()` call.

It seems to be relying on timing, as the test consistently takes 500ms on my machine, when it should take much less than that, or at least vary in timing. I saw this in the logs once:

```
time="2022-07-28T12:00:58+02:00" level=debug category="Browser:Close" elapsed="500 ms" goroutine=19
```
... but don't see it consistently, since that elapsed measurement happens by chance from the last logger call.

I don't see where the delay takes place... We set `SlowMo` to 0, and changing [`LifeCycleNetworkIdleTimeout`](https://github.com/grafana/xk6-browser/blob/1997670b94d3252768c2a5da80b645957634be3b/common/consts.go#L35) didn't have an effect. :confused:

So I think we should make sure that this test doesn't rely on timing, and that these Promises are actually resolved and waited on.

Reply:
Yeah, same behaviour when i ran it locally. Although the logic seems to be fine, just need to work with `resolve`/`reject`.

We should also test that `reject` is being called when the `javascript` run fails for whatever reason (maybe it times out or doesn't find the element).

Reply:
I've amended this test so now it's checking the state of the promise before ending the test. This should be enough to show that the promises are resolving. What isn't clear is why we don't need `WaitForNavigation` i.e. I can remove it and the navigation occurs and the second page loads before we assert the title. Does `Click` inherently wait for the navigation to complete?

Reply:
> Does `Click` inherently wait for the navigation to complete?

No, but it's likely that the navigation happens very quickly, so that by the time `tb.await()` returns, the promises have already been fulfilled.

You can confirm this by intercepting the click on the element, and delaying the amount of time for the navigation to happen.

For example, change the `/first` body to:

```html
<body>
  <a href="/second">click me</a>
  <script>
    const el = document.querySelector('a');
    el.addEventListener('click', function(evt) {
      evt.preventDefault();
      setTimeout(() => {
        window.location = '/second';
      }, 1000);
    });
  </script>
</body>
```

Then if you remove the `p.WaitForNavigation()` call in `tb.await()`, the test will fail with:

```
Error:          Not equal:
                expected: "Second page"
                actual  : "First page"
```

And if you add it back, it will pass.

BTW, we should probably set a timeout for `WaitForNavigation` like you did in the `WaitForFrameNavigationWithinDocument` test, otherwise if something goes wrong we'll have to wait the default 30s. If we delay the click by 1s, then timing out at 5s for the navigation should be enough, even on CI machines.

Reply:
Resolved in: 1e9fda0a58e452bdb611b9a61f5617bdb7e8f60d

NOTE: I didn't add the one second delay. I don't understand why we would need that, and I'm hoping we don't need it 🤞

Reply:
Sure, it's not strictly needed, but it would somewhat ensure that `WaitForNavigation()` is also triggered on fast machines.

But it's not a big deal, let's leave it as is.